test/e2e/xfn_test.go

```diff
@@ -0,0 +1,318 @@
+		} else {
+			t.Fatalf("not a map[string]string: %v type %s", v, reflect.TypeOf(v))
+		}
+	}
+	return s
+}
+
+func asMapOfStrings(m map[string]interface{}) map[string]string {
+	r := make(map[string]string)
+	for k, v := range m {
+		r[k] = fmt.Sprintf("%v", v)
+	}
+	return r
+}
+
+// ManagedResourcedProcessedByFunction asserts that MRs contains the requested label
+func ManagedResourcedProcessedByFunction() features.Func {
+
+	return func(ctx context.Context, t *testing.T, config *envconf.Config) context.Context {
+		labelName := "labelizer.xfn.crossplane.io/processed"
+		rg := resourceGetter(ctx, t, config)
+		claim := rg("fn-labelizer", "default", "nop.example.org/v1alpha1", "NopResource")
+		r := resourceValue(t, claim, "spec", "resourceRef")
+
+		xr := rg(r["name"], "default", r["apiVersion"], r["kind"])
+		mrefs := resourceSliceValue(t, xr, "spec", "resourceRefs")
+		for _, mref := range mrefs {
+			err := wait.For(func() (done bool, err error) {
+				mr := rg(mref["name"], "default", mref["apiVersion"], mref["kind"])
+				l, found := mr.GetLabels()[labelName]
+				if !found {
+					return false, nil
+				}
+				if l != "true" {
+					return false, nil
+				}
+				return true, nil
+			}, wait.WithTimeout(5*time.Minute))
+			if err != nil {
+				t.Fatalf("Expected label %v value to be true", labelName)
+			}
+
+		}
+		return ctx
+	}
+}
+
+// CreateTLSCertificateAsSecret for given dns name and store the secret in the given namespace
+func CreateTLSCertificateAsSecret(dnsName string, ns string) features.Func {
+	return func(ctx context.Context, t *testing.T, config *envconf.Config) context.Context {
+		caPem, keyPem, err := createCert(dnsName)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		secret := &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "reg-cert",
+				Namespace: ns,
+			},
+			Type: corev1.SecretTypeTLS,
+			StringData: map[string]string{
+				"tls.crt": caPem,
+				"tls.key": keyPem,
+			},
+		}
+		client := config.Client().Resources()
+		if err := client.Create(ctx, secret); err != nil {
+			t.Fatalf("Cannot create secret %s: %v", secret.Name, err)
+		}
+		configMap := &corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "reg-ca",
+				Namespace: namespace,
+			},
+			Data: map[string]string{
+				"domain.crt": caPem,
+			},
+		}
+		if err := client.Create(ctx, configMap); err != nil {
+			t.Fatalf("Cannot create config %s: %v", configMap.Name, err)
+		}
+		return ctx
+	}
+}
+
+// CopyImagesToRegistry copies fn images to private registry
+func CopyImagesToRegistry() features.Func {
+	return func(ctx context.Context, t *testing.T, config *envconf.Config) context.Context {
+		nodes := &corev1.NodeList{}
+		if err := config.Client().Resources().List(ctx, nodes); err != nil {
+			t.Fatal("cannot list nodes", err)
+		}
+		if len(nodes.Items) == 0 {
+			t.Fatalf("no nodes in the cluster")
+		}
+		i := slices.IndexFunc(nodes.Items[0].Status.Addresses, func(a corev1.NodeAddress) bool {
```

Comment:
Given that this is the only reason we are adding `golang.org/x/exp` as a dependency, which although nice, is still untagged and therefore would result in noisy version bumps every month (according to the schedule we've set for go dependencies referenced by digest), I'd say here we can just switch to a simple for loop.